### ISSUE TITLE
Usage with Tornado (was: No DB associated to model)

### ISSUE THREAD

**Issue by asyncins** [tortoise/tortoise-orm#99]
```
from tortoise.models import Model
from tortoise.fields import *

class Tournament(Model):
    id = IntField(pk=True)
    name = TextField()

if __name__ == "__main__":
    from tortoise import Tortoise, run_async


    async def init():
        # Here we create a SQLite DB using file "db.sqlite3"
        #  also specify the app name of "models"
        #  which contain models from "app.models"
        await Tortoise.init(
            db_url='sqlite://otc.db',
            modules={'models': ['model']} # 传入模块名称，本模块名为 model
        )
        # Generate the schema
        await Tortoise.generate_schemas()


    run_async(init())

    async def dps():
        await Tournament.create(name='Another Tournament')

        # Now search for a record
        tour = await Tournament.filter(name__contains='Another').first()
        print(tour.name)
    run_async(dps())
```
==============================
tortoise.exceptions.ConfigurationError: No DB associated to model

WHY ?

**Comment by grigi** [comment]
Hi asyncins

`run_async()` is a helper function that ensures connections are closed when it is done:
https://tortoise-orm.readthedocs.io/en/latest/setup.html#cleaningup
https://tortoise-orm.readthedocs.io/en/latest/setup.html#tortoise.run_async

Only use it once:
```py
from tortoise import Tortoise, run_async
from tortoise.models import Model
from tortoise.fields import *

class Tournament(Model):
    id = IntField(pk=True)
    name = TextField()

async def init():
    # Here we create a SQLite DB using file "db.sqlite3"
    #  also specify the app name of "models"
    #  which contain models from "app.models"
    await Tortoise.init(
        db_url='sqlite://otc.db',
        modules={'models': ['model']} # 传入模块名称，本模块名为 model
    )
    # Generate the schema
    await Tortoise.generate_schemas()

async def dps():
    await Tournament.create(name='Another Tournament')

    # Now search for a record
    tour = await Tournament.filter(name__contains='Another').first()
    print(tour.name)

async def main():
    await init()
    await dps()

if __name__ == "__main__":
    run_async(main())
```

Or, alternatively manage the loop & cleanup yourself.
```py
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(main())
    finally:
        loop.run_until_complete(Tortoise.close_connections())
```

**Comment by asyncins** [comment]
Traceback (most recent call last):
  File "/home/asyncins/anaconda3/envs/envpoll/lib/python3.7/site-packages/tortoise/models.py", line 47, in db
    return current_transaction_map[self.default_connection].get()
KeyError: None

**Comment by grigi** [comment]
Sorry, I didn't test my example, there was 2 issues.
In `Tortoise.init()`, the modules param, it tries to import a module called "models" which would work if you have a models.py in the path.
If you define the models in the same file, you should either use the name of the module, or `'__main__'`
When I take that example and do:
```
    await Tortoise.init(
        db_url='sqlite://otc.db',
        modules={'models': ['__main__']} # Change here
    )
    # Generate the schema
```
It works for me

**Comment by grigi** [comment]
I don't have much experience with Tornado, at least not since they adopted `asyncio` event loop.
I imagine it will have a setup section, and in there you call `await init()`, and in the teardown section you should call `await Tortoise.close_connections()`.
Then in the HTTP requests you should just use the ORM as per normal.

We should probably have a small tutorial or integration projects for the common frameworks. :thinking:

**Comment by asyncins** [comment]
> I don't have much experience with Tornado, at least not since they adopted `asyncio` event loop.
> I imagine it will have a setup section, and in there you call `await init()`, and in the teardown section you should call `await Tortoise.close_connections()`.
> Then in the HTTP requests you should just use the ORM as per normal.
> 
> We should probably have a small tutorial or integration projects for the common frameworks. thinking

I like this ORM very much. It's very similar to Django  ORM. It's very enjoyable to use.

I now have a lightweight Tornado project that wants to use the same lightweight Sqlite, but now it seems that they don't work well together, which is a great pity.

**Comment by SultanNasyrovDeveloper** [comment]
Hello everyone! 
I get the same error in my aiohttp project.

Here is my db_init.py

    from tortoise import Tortoise, run_async
    from tight_trader.settings import config


    DSN = "postgres://{user}:{password}@{host}:{port}/{database}"


    async def create_tables(url):
        await Tortoise.init(
            db_url=url, modules={'models': ['tight_trader.trading_bot.models',
                                        ]}
        )
        await Tortoise.generate_schemas()


    if __name__ == '__main__':
        db_url = DSN.format(**config['postgres'])
        run_async(create_tables(db_url))


models.py

    import asyncio

    from aiologger import Logger
    from tortoise.models import Model
    from tortoise import fields

    from tight_trader.trading_bot.enums import BotType


    logger = Logger.with_default_handlers(name=__name__)


    class BaseTradingBot(Model):
        """ Abstract trading bot model """
        id = fields.IntField(pk=True)
        name = fields.CharField(max_length=100)
        type = None

        on = fields.BooleanField(default=False)
        frequency = fields.IntField(default=1)

        instruments = None  # to set up all instruments bot can trade

        class Meta:
            abstract = True

    def __str__(self):
        return 'Trading Bot {}'.format(self.id)



    def _tear_down(self):
        """ Makes all actions bot needs to stop trading"""
        self.session.stop()



    class CryptoBot(BaseTradingBot):
        """  """
        type = BotType.CRYPTO


I get the error when i try to 

    @aiohttp_jinja2.template('trading_bot/list.html')
    def bot_list(request):
         context = {'bots': CryptoBot.all()}

**Comment by SultanNasyrovDeveloper** [comment]
I didn't understand what is the right way to solve the problem

**Comment by abondar** [comment]
@SultanNasyrovDeveloper could you please show stacktrace for error and elaborate on how are you launching you script

**Comment by ozturkberkay** [comment]
Tortoise works (+ with migrations) in Tornado 6.0.3 and I have been using it for a while:

Define the models inside a folder called `models` under the root directory of the project. Just create an `__init__.py` file like this:
```python
__models__ = [
    SomeModel,
    ...
]
```

Then use `tornado.options` for your configurations:
```python
{
    'connections': {
         'mysql': {
            'engine': 'tortoise.backends.mysql',
            'credentials': {
                'host': options.mysql_host,
                'port': options.mysql_port,
                'user': options.mysql_user,
                'password': options.mysql_password,
                'database': options.mysql_database,
		},
        }
    },
    'apps': {
	'models': {'models': ['models'], 'default_connection': 'mysql'}
    },
}
```

For your `tornado.web.Application`,  you can create your entry-point like this:
```python
try:
    IOLoop.current().run_sync(init)  # Initialize your database here.
    IOLoop.current().start()  # Application starts here.
except KeyboardInterrupt:
    _logger.warning('Keyboard interruption detected!')
finally:
    IOLoop.current().stop()
    _logger.warning('Closing the database connections!')
   # Using Tornado's IOLoop throws this error for some reason:
   # tornado.util.TimeoutError: Operation timed out after None seconds
   # So, I use asyncio here.
   get_event_loop().run_until_complete(Tortoise.close_connections())
```

You can then do the initialization procedures you want inside the `init` function:
```python
await Tortoise.init(...)

if options.mysql_sync:
    # Generate schemas
    # Init table entries
    # ...
```

**Comment by grigi** [comment]
Thanks for this, Will help in building an example/docs.

**Comment by keli5** [comment]
I have the same problem: I have initdb.py which works fine:
```
from tortoise import Tortoise, run_async

async def init():
    await Tortoise.init(
        db_url='sqlite://lettersbot_data.sqlite3',
        modules={'models': ["models"]}
    )
    # Generate the schema
    await Tortoise.generate_schemas()

run_async(init())
```
Models.py:
```
from tortoise.models import Model
from tortoise import fields

class LBUser(Model):
    balance = fields.IntField()
    canUseBot = fields.BooleanField()
    inventory = fields.JSONField()

class LBGuild(Model):
    muteRole = fields.IntField()
```
But when I try to use `get_or_create`, it errors:
```
from tortoise.models import Model
from models import LBUser, LBGuild
async def on_message(self, message):
    user = await LBUser.get_or_create(
        "balance": 0,
        "canUseBot": True,
        "inventory": []
    })
```
When on_message is fired I get an error:
`tortoise.exceptions.ConfigurationError: No DB associated to model`

**Comment by misss85246** [comment]
Hi there, I got the same error in sanic 
this is my Folder tree:
```text
.
├── docs
│   └── test.md
├── src
│   ├── config.py
│   ├── __init__.py
│   ├── logs
│   │   └── blog.log
│   ├── models.py
│   ├── server.py
│   ├── tools
│   │   ├── __init__.py
│   │   ├── loggerFormatTools.py
│   │   ├── responseFormatTools.py
│   │   └── simpleTools.py
│   ├── urls.py
│   └── views
│       ├── blogViews.py
│       └── __init__.py
└── test
    ├── aaa.py
    ├── dbClientTest.py
    ├── functionTest.py
    ├── __init__.py
    └── markdownTest.py
```
there are some model in `src/models.py`
```python
class Blog(models.Model):
    id = fields.IntField(pk=True)
    aritcle_id = fields.CharField(max_length=20, unique=True)
    titile = fields.CharField(max_length=128)
    digest = fields.CharField(max_length=255)
    cover = fields.CharField(max_length=255, null=True)
    content = fields.TextField()
    tags = fields.CharField(max_length=255)
    browse = fields.IntField(default=0)
    praise = fields.IntField(default=0)
    c_time = fields.DatetimeField(auto_now_add=True)
    u_time = fields.DatetimeField(auto_now=True)
    is_delete = fields.BooleanField(default=False)

    def __str__(self):
        return f"<BlogObj article_id={self.aritcle_id}>"

    class Meta:
        table = 'blogs'
        index = ["tags", "id_delete"]
        ordering = ["-c_time", "-u_time"]
```
and my config file like this: 
```
TORTOISE_CONFIG = {
    "db_url": "mysql://admin:admin@127.0.0.1:3306/Test?maxsize=32&minsize=8&echo=True",
    "modules": {
        'models': ['models'],
    }
}
```
I tried got blogs in `blogView.py`,  but It raise error
```python
No DB associated to model
``` 
I have tried to create a models folder, and then move `models.py` into and create the `__init__.py` file.
And I change the TORTOISE_CONFIG  after that:
``` python
TORTOISE_CONFIG = {
    "db_url": "mysql://admin:admin@127.0.0.1:3306/Test?maxsize=32&minsize=8&echo=True",
    "modules": {
        'models': ['__main__'],
    }
}
```
It raise another error :
``` python
RuntimeWarning: Module "__main__" has no models
  cls._init_apps(apps_config)
```
what should I do ?

**Comment by long2ice** [comment]
Run `await Tortoise.init()` on app startup.

**Comment by misss85246** [comment]
> Run `await Tortoise.init()` on app startup.

I use this in `app.py` : 

```python
register_tortoise(server, **server.config.TORTOISE_CONFIG)
```
if I use `Tortoise.init()` How should I pass the parameter *app* ?
can you give me a simple example ?

**Comment by long2ice** [comment]
see here https://tortoise-orm.readthedocs.io/en/latest/examples/fastapi.html#

**Comment by misss85246** [comment]
@long2ice 

Hi, Dear, I think maybe I know how to use it
If you want to use tortoise-orm, you must refer to all the models in the registration page of tortoise.
If you use the configuration of' __main__', you need to manually reference the model
If you are using the path of the model file, it needs to be a relative path relative to the registered tortoise file
Like in my question:
My file tree is:
```text
.
├── docs
│   └── test.md
├── src
│   ├── config.py
│   ├── __init__.py
│   ├── logs
│   │   └── blog.log
│   ├── models.py
│   ├── server.py
│   ├── tools
│   │   ├── __init__.py
│   │   ├── loggerFormatTools.py
│   │   ├── responseFormatTools.py
│   │   └── simpleTools.py
│   ├── urls.py
│   └── views
│       ├── blogViews.py
│       └── __init__.py
└── test
    ├── aaa.py
    ├── dbClientTest.py
    ├── functionTest.py
    ├── __init__.py
    └── markdownTest.py
```
I registered with the file src/server.py and used tortoise-orm, like this:
```python
from sanic import Sanic
from src.config import PROJECT_CONFIG, LOGGING_CONFIG, SERVER_CONFIG
from src.urls import server_bp
from tortoise.contrib.sanic import register_tortoise

server = Sanic('abcdefg', log_config=LOGGING_CONFIG)
server.config.update(PROJECT_CONFIG)
server.blueprint(server_bp)

register_tortoise(server, **server.config.TORTOISE_CONFIG)
if __name__ == '__main__':
    print(server.config.BASE_DIR)
    server.run(**SERVER_CONFIG)
```
Then I need to add the path of model.py relative to the server file in the configuration file, that is, src.model':
```python
TORTOISE_CONFIG = {
    "db_url": "mysql://admin:admin@127.0.0.1:3306/Blog?maxsize=32&minsize=8&echo=True",
    "modules": {
        'models': ['src.models'],
    }
}
```
If I use `__main__` in `models`, like this:
```python
TORTOISE_CONFIG = {
    "db_url": "mysql://admin:admin@127.0.0.1:3306/Blog?maxsize=32&minsize=8&echo=True",
    "modules": {
        'models': ['__main__'],
    }
}
```
I need import all models in `server.py`, like this:
```python
from sanic import Sanic
from src.config import PROJECT_CONFIG, LOGGING_CONFIG, SERVER_CONFIG
from src.urls import server_bp
from tortoise.contrib.sanic import register_tortoise
from src.models import *

server = Sanic('abcdefg', log_config=LOGGING_CONFIG)
server.config.update(PROJECT_CONFIG)
server.blueprint(server_bp)

register_tortoise(server, **server.config.TORTOISE_CONFIG)
if __name__ == '__main__':
    print(server.config.BASE_DIR)
    server.run(**SERVER_CONFIG)
```
It worked very well, I think other frameworks are the same principle, 
If you get same error, please check your model path in your config
Hope It's useful to you !
